doc/design/error_reporting.md

```diff
@@ -0,0 +1,456 @@
+
+In this way, the custom implementation defines a contract for the errors.
+
+### Reponsibilty of a Module
+
+1. Define all its errors and error codes
+1. The errors must satisfy the error contract conditions required by the custom implementation.
+1. Define how codes are converted to errors
+1. Combine the definitions with the API and provide a header that defines its whole error reporting.
+
+Once this is defined, the module can use the error reporting in its own functions.
+
+### Kinds of Errors
+
+There are three mandatory kinds (or categories) of errors.
+
+1. Fatal errors
+1. Precondition violations
+1. Assumption violations
+
+All of them must be supported by the custom implementation and all of them abort execution (which cannot be
+changed).
+
+Furthermore the custom implementation may define additional kinds of errors, specifically non-fatal errors.
+For non-fatal errors the execution continues afer reporting, so they should be used when the
+circumstances allow recovery from the error.
+
+### Panic
+
+Before execution is aborted, a special `panic` handler is invoked. The custom implementation defines this handler
+as part of the backend.
+
+## Using the API to Report Errors
+
+Assume the module defines some error codes
+
+```cpp
+enum class Code
+{    
+    OutOfMemory = 73,
+    OutOfBounds = 21
+};
+```
+
+and the custom implementation defines a non-fatal `RUNTIME_ERROR`.
+
+Then we may use the following functions
+
+### Panic
+
+```cpp
+IOX_PANIC("some message");
+
+IOX_PANIC("");
+```
+
+Signals panic, invokes the panic handler and aborts execution.
+
+Even if there is no message to be provided, it has to be called with an empty message. This is a
+technical limitation from macro usage (required for location) as it is not allowed to have empty variadic
+macros.
+
+### Report an Error
+
+To report an error the error code (later to be extended to error objects) has to be provided
+
+```cpp
+IOX_REPORT(Code::OutOfMemory, RUNTIME_ERROR);
+```
+
+This reports a `OutOfMemory` error and continues execution.
+
+### Report a Fatal Error
+
+Similarly
+
+```cpp
+IOX_REPORT_FATAL(Code::OutOfMemory);
+```
+
+reports a fatal error that aborts execution after the custom implementation specific handler is invoked.
+
+Decoupling the error and its category is intentional, as e.g. an `OutOfMemory` error may not always be
+fatal. It may become fatal after it is propagated further along the call stack. 
+Propagation is possible by various means, e.g. return codes, monadic types or even exceptions (that
+must be caught before reporting the error elsewhere again).
+
+### Conditionally Report an Error
+
+Conditionally reporting an error if some condition holds is useful for more compact error reporting.
+
+```cpp
+int x;
+// ...
+IOX_REPORT_IF(x<0, Code::OutOfBounds, RUNTIME_ERROR)
+```
+
+### Assert That a Condition Holds
+
+Similarly we can conditionally check whether a condition does not hold and report a fatal error in
+the case that it does not
```

Comment:
```suggestion
Similarly we can conditionally check whether a condition does hold and report a fatal error in
the case that it does not
```
Otherwise it is like a double negation.